File: lib/services/chat_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/supabase_client.dart';
import '../models/app_user.dart';

/// Port of src/services/chat.service.ts.

class ChatMessage {
  const ChatMessage({required this.role, required this.text, this.time});

  /// 'user' or 'ai', matching the chat_role enum.
  final String role;
  final String text;
  final String? time;

  bool get isUser => role == 'user';

  Map<String, dynamic> toWire() => {'role': role, 'text': text};
}

class SendMessageResult {
  const SendMessageResult({
    required this.reply,
    required this.chatCount,
    required this.limit,
    required this.isSubscribed,
  });

  final String reply;
  final int chatCount;
  final int limit;
  final bool isSubscribed;
}

/// Raised when the free-chat allowance is exhausted, so the UI can show the
/// subscribe prompt rather than a generic error.
class ChatLimitException implements Exception {
  const ChatLimitException();

  @override
  String toString() => 'LIMIT_REACHED';
}

class ChatService {
  const ChatService();

  Future<SendMessageResult> send({
    required String message,
    required List<ChatMessage> history,
    required AppUser? user,
  }) async {
    if (currentUserId == null) {
      throw const ApiException('You must be signed in to use chat.');
    }

    // Pre-flight check against the local profile (fast, avoids a round-trip).
    // The edge function re-checks authoritatively — this is convenience, not
    // the limit itself.
    if (user != null && !user.isSubscribed && user.chatLimitReached) {
      throw const ChatLimitException();
    }

    final token = currentAccessToken;
    final response = await http.post(
      Uri.parse('${SupabaseConfig.resolvedFunctionsBaseUrl}/chat/send'),
      headers: {
        'Content-Type': 'application/json',
        if (token != null) 'Authorization': 'Bearer $token',
      },
      body: jsonEncode({
        'message': message,
        'history': history.map((m) => m.toWire()).toList(),
      }),
    );

    final body = _decode(response.body);

    if (response.statusCode != 200) {
      if (response.statusCode == 403 && body?['limit_reached'] == true) {
        throw const ChatLimitException();
      }
      throw ApiException(
        body?['message'] as String? ?? 'Chat request failed.',
      );
    }

    return SendMessageResult(
      reply: body?['reply'] as String? ?? '',
      chatCount: (body?['chat_count'] as num?)?.toInt() ?? 0,
      limit: (body?['limit'] as num?)?.toInt() ?? freeChatLimit,
      isSubscribed: body?['is_subscribed'] as bool? ?? false,
    );
  }

  /// Chat history.
  ///
  /// chat_messages_select scopes this to the caller's own rows, so the
  /// user_id filter is belt-and-braces rather than the security boundary.
  Future<List<ChatMessage>> history() async {
    final userId = currentUserId;
    if (userId == null) {
      throw const ApiException('You must be signed in to use chat.');
    }

    final rows = await supabase
        .from('chat_messages')
        .select('role, text, created_at')
        .eq('user_id', userId)
        .order('created_at', ascending: true)
        .limit(100);

    return rows
        .map(
          (row) => ChatMessage(
            role: row['role'] as String,
            text: row['text'] as String? ?? '',
            time: _formatTime(row['created_at'] as String?),
          ),
        )
        .toList();
  }

  static Map<String, dynamic>? _decode(String body) {
    try {
      final decoded = jsonDecode(body);
      return decoded is Map<String, dynamic> ? decoded : null;
    } on FormatException {
      return null;
    }
  }

  static String? _formatTime(String? iso) {
    if (iso == null) return null;
    final parsed = DateTime.tryParse(iso);
    if (parsed == null) return null;
    final local = parsed.toLocal();
    final hour = local.hour.toString().padLeft(2, '0');
    final minute = local.minute.toString().padLeft(2, '0');
    return '$hour:$minute';
  }
}

File: src/services/chat.service.ts
import { supabase, functionsBaseUrl, getAccessToken, getCurrentUserId } from "./supabase";
import {
  FREE_CHAT_LIMIT,
  canUseFreeChats,
  isSubscriptionActive,
} from "./subscription.model";

export interface ChatMessage {
  role: "user" | "ai";
  text: string;
  time?: string;
}

export interface SendMessageResponse {
  reply: string;
  chat_count: number;
  limit: number;
  is_subscribed: boolean;
}

export interface ChatStatusResponse {
  chat_count: number;
  limit: number;
  is_subscribed: boolean;
  free_chat_eligible: boolean;
  limit_reached: boolean;
  remaining: number;
}

export interface ChatHistoryResponse {
  messages: ChatMessage[];
}

export class ChatLimitError extends Error {
  constructor() {
    super("LIMIT_REACHED");
    this.name = "ChatLimitError";
  }
}

// ── Public API ───────────────────────────────────────────────────────────────

const requireUserId = async () => {
  const id = await getCurrentUserId();
  if (!id) throw new Error("You must be signed in to use chat.");
  return id;
};

export const getChatStatus = async (
  token: string | null,
): Promise<ChatStatusResponse> => {
  void token;
  await requireUserId();

  const { getCurrentUserProfile } = await import("./auth.service");
  const user = await getCurrentUserProfile();
  const chatCount = user?.chat_count ?? 0;
  const isSubscribed = isSubscriptionActive(user);
  const freeChatEligible = canUseFreeChats(user);

  return {
    chat_count: chatCount,
    limit: FREE_CHAT_LIMIT,
    is_subscribed: isSubscribed,
    free_chat_eligible: freeChatEligible,
    limit_reached: !isSubscribed && (!freeChatEligible || chatCount >= FREE_CHAT_LIMIT),
    remaining: isSubscribed || freeChatEligible
      ? Math.max(0, FREE_CHAT_LIMIT - chatCount)
      : 0,
  };
};

export const sendMessage = async (
  message: string,
  token: string | null,
  history: ChatMessage[] = [],
): Promise<SendMessageResponse> => {
  void token;
  await requireUserId();

  // Pre-flight limit check using the local profile (fast, avoids a
  // round-trip). The edge function re-checks authoritatively.
  const status = await getChatStatus(null);
  if (!status.is_subscribed && status.limit_reached) {
    throw new ChatLimitError();
  }

  // The dual Firebase-Functions / Supabase-Functions branch is gone: there is
  // only one backend now. The OpenAI key stays server-side either way.
  const accessToken = await getAccessToken();
  const response = await fetch(`${functionsBaseUrl}/chat/send`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
    },
    body: JSON.stringify({ message, history }),
  });

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    if (response.status === 403 && data?.limit_reached) throw new ChatLimitError();
    throw new Error(data?.message ?? "Chat request failed.");
  }

  return data as SendMessageResponse;
};

const formatMessageTime = (value: string | null): string => {
  if (!value) return "";
  const date = new Date(value);
  return Number.isNaN(date.getTime())
    ? ""
    : date.toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" });
};

export const getChatHistory = async (
  token: string | null,
): Promise<ChatHistoryResponse> => {
  void token;
  const userId = await requireUserId();

  // chat_messages_select scopes this to the caller's own rows, so the
  // user_id filter is belt-and-braces rather than the security boundary.
  const { data, error } = await supabase
    .from("chat_messages")
    .select("role, text, created_at")
    .eq("user_id", userId)
    .order("created_at", { ascending: true })
    .limit(100);

  if (error) throw new Error(error.message);

  return {
    messages: (data ?? []).map((item) => ({
      role: item.role,
      text: item.text,
      time: formatMessageTime(item.created_at),
    })) as ChatMessage[],
  };
};
